compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2481,6 +2481,15 @@ impl<'a> Parser<'a> {
 
         let pat = self.recover_parens_around_for_head(pat, begin_paren);
 
+        // Recover from missing expression in `for` loop
+        if matches!(expr.kind, ExprKind::Block(..))
+            && !matches!(self.token.kind, token::OpenDelim(token::Delimiter::Brace))
+            && self.may_recover()
+        {
+            self.sess.emit_err(MissingExpressionInForLoop { span: expr.span });
+            return Ok(self.mk_expr(expr.span, ExprKind::Err));
+        }
```

Comment:
One last suggestion: We should be constructing a *for loop* here. Do something approximately like this--

```suggestion
        if let ExprKind::Block(block, label) = expr.kind
            && !matches!(self.token.kind, token::OpenDelim(token::Delimiter::Brace))
            && self.may_recover()
        {
            self.sess.emit_err(MissingExpressionInForLoop { span: expr.span });
            let err_expr = self.mk_expr(expr.span, ExprKind::Err);
            return Ok(self.mk_expr(lo.to(self.prev_token.span), ExprKind::ForLoop(pat, err_expr, block, label)));
        }
```

You *might* need to clone some stuff, I haven't checked if it compiles.